src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+        /// </summary>
+        private static int GetInsertionIndex(Compilation compilation, IMethodSymbol methodSymbol,
+            ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
+            Contract.ThrowIfNull(parameterList);
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in methodSymbol.Parameters)
+            {
+                // Want to skip optional parameters, params parameters, and CancellationToken since they should be at
+                // the end of the list.
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams &&
+                        !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// For the trampoline case, it goes through the invocations and adds an argument which is a 
+        /// call to the extracted method.
+        /// Introduces a new method overload or new trampoline method.
+        /// Updates the original method site with a newly introduced parameter.
+        /// </summary>
+        private static async Task<Solution> ModifyDocumentInvocationsTrampolineAndIntroduceParameterAsync(
+            Compilation compilation, Solution modifiedSolution, Document currentDocument, Document originalDocument,
+            ImmutableArray<TInvocationExpressionSyntax> invocations, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            bool allOccurrences, string parameterName, TExpressionSyntax expression, bool trampoline, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(currentDocument);
+            var semanticFacts = currentDocument.GetRequiredLanguageService<ISemanticFactsService>();
+            var invocationSemanticModel = await currentDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var root = await currentDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = currentDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+
+            if (trampoline)
+            {
+                foreach (var invocationExpression in invocations)
+                {
+                    editor.ReplaceNode(invocationExpression, (currentInvocation, _) =>
+                    {
+                        return GenerateNewInvocationExpressionForTrampoline(syntaxFacts, editor, currentInvocation,
```

Comment:
passing in the editor here is scary.

Reply:
I technically don't need to, but can you explain why it would be scary?

Reply:
Because in the callback where the editor is computing what it needs to mutate the result, you are paying in the thing that can be mutated more. It's like mutating a list as your iterating over it. It is very unclear what it means.

Reply:
right, I completely missed that it is in the editor's callback. makes sense